tests/test_datasets/test_hierarchical_structure.py

```diff
@@ -0,0 +1,176 @@
+    return HierarchicalStructure(
+        level_structure={"total": ["X", "Y"], "X": ["a"], "Y": ["b"], "a": ["c"], "b": ["d"]},
+        level_names=["l1", "l2", "l3", "l4"],
+    )
+
+
+@pytest.mark.parametrize(
+    "target,source",
+    (
+        ("l1", "l2"),
+        ("l2", "l3"),
+    ),
+)
+def test_get_summing_matrix_format(simple_hierarchical_struct: HierarchicalStructure, source: str, target: str):
+    output = simple_hierarchical_struct.get_summing_matrix(target_level=target, source_level=source)
+    assert isinstance(output, scipy.sparse.base.spmatrix)
+
+
+@pytest.mark.parametrize(
+    "struct, target,source,answer",
+    (
+        ("tailed_hierarchical_struct", "l1", "l2", np.array([[1, 1]])),
+        ("tailed_hierarchical_struct", "l1", "l3", np.array([[1, 1, 1]])),
+        ("tailed_hierarchical_struct", "l1", "l4", np.array([[1, 1, 1, 1]])),
+        ("tailed_hierarchical_struct", "l2", "l3", np.array([[1, 0, 0], [0, 1, 1]])),
+        ("tailed_hierarchical_struct", "l2", "l4", np.array([[1, 1, 0, 0], [0, 0, 1, 1]])),
+        ("tailed_hierarchical_struct", "l3", "l4", np.array([[1, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]])),
+        ("simple_hierarchical_struct", "l1", "l2", np.array([[1, 1]])),
+        ("simple_hierarchical_struct", "l1", "l3", np.array([[1, 1, 1, 1]])),
+        ("simple_hierarchical_struct", "l2", "l3", np.array([[1, 1, 0, 0], [0, 0, 1, 1]])),
+        ("long_hierarchical_struct", "l1", "l2", np.array([[1, 1]])),
+        ("long_hierarchical_struct", "l1", "l3", np.array([[1, 1]])),
+        ("long_hierarchical_struct", "l1", "l4", np.array([[1, 1]])),
+        ("long_hierarchical_struct", "l2", "l3", np.array([[1, 0], [0, 1]])),
+        ("long_hierarchical_struct", "l2", "l4", np.array([[1, 0], [0, 1]])),
+        ("long_hierarchical_struct", "l3", "l4", np.array([[1, 0], [0, 1]])),
+    ),
+)
+def test_summing_matrix(struct: str, source: str, target: str, answer: np.ndarray, request: pytest.FixtureRequest):
+    np.testing.assert_array_almost_equal(
+        answer, request.getfixturevalue(struct).get_summing_matrix(target_level=target, source_level=source).toarray()
+    )
+
+
+@pytest.mark.parametrize(
+    "target,source,error",
+    (
+        ("l0", "l2", "Invalid level name: l0"),
+        ("l1", "l0", "Invalid level name: l0"),
+        ("l2", "l1", "Target level must be higher in hierarchy than source level!"),
+    ),
+)
+def test_level_transition_errors(
+    simple_hierarchical_struct: HierarchicalStructure,
+    target: str,
+    source: str,
+    error: str,
+):
+    with pytest.raises(ValueError, match=error):
+        simple_hierarchical_struct.get_summing_matrix(target_level=target, source_level=source)
+
+
+@pytest.mark.parametrize(
+    "structure,answer",
+    (
+        ({"total": ["X", "Y"], "X": ["a", "b"], "Y": ["c", "d"]}, "total"),
+        ({"X": ["a", "b"]}, "X"),
+    ),
+)
+def test_root_finding(structure: Dict[str, List[str]], answer: str):
```

Comment:
You also need tests to check the correctness of defining the other attributes like `_num_nodes `, `_hierarchy_leaves `